runtime/src/main/java/org/corfudb/runtime/collections/streaming/StreamPollingScheduler.java

```diff
@@ -316,32 +321,52 @@ private void handleFailures(List<StreamingTask> failedTasks) {
-
-        final long cycleStartTimeNs = System.nanoTime();
-
-        final List<StreamingTask> runnableTasks = getTasks(StreamStatus.RUNNABLE);
-        final List<StreamingTask> syncingTasks = getTasks(StreamStatus.SYNCING,
-                t -> t.getStream().availableSpace() >= pollThreshold);
-        final List<StreamingTask> failedTasks = getTasks(StreamStatus.ERROR);
-
-        runnableTasks.forEach(t -> t.move(StreamStatus.RUNNABLE, StreamStatus.SCHEDULING));
-        poll(runnableTasks);
-        dispatchSyncTasks(runnableTasks);
-
-        // Since these tasks are already syncing, they shouldn't be dispatched again
-        poll(syncingTasks);
-
-        // handle tasks that have failed (i.e., remove them and propagate the errors to the listeners)
-        handleFailures(failedTasks);
-
-        long elapsedTimeNs = System.nanoTime() - cycleStartTimeNs;
-        MicroMeterUtils.time(Duration.ofNanos(elapsedTimeNs), "schedule.cycle");
-        long nextSchedule = Math.max(0, pollPeriod.toNanos() - elapsedTimeNs);
-        Preconditions.checkState(nextSchedule <= pollPeriod.toNanos());
-
-        scheduler.schedule(this::tick, nextSchedule, TimeUnit.NANOSECONDS);
+        try {
+            // Block thread until there's work to do
+            waitForTasks();
+
+            final long cycleStartTimeNs = System.nanoTime();
+
+            final List<StreamingTask> runnableTasks = getTasks(StreamStatus.RUNNABLE);
+            final List<StreamingTask> syncingTasks = getTasks(StreamStatus.SYNCING,
+                    t -> t.getStream().availableSpace() >= pollThreshold);
+            final List<StreamingTask> failedTasks = getTasks(StreamStatus.ERROR);
+
+            runnableTasks.forEach(t -> {
+                try {
+                    t.move(StreamStatus.RUNNABLE, StreamStatus.SCHEDULING);
+                } catch (Exception e) {
+                    t.setError(e);
+                    log.error("StreamingPollingScheduler: encountered exception {} while moving to 'scheduling' status, " +
+                            "listener={} with id={} onError()", e, t.getListener(), t.getListenerId());
+                }
+            });
+            poll(runnableTasks);
+            dispatchSyncTasks(runnableTasks);
+
+            // Since these tasks are already syncing, they shouldn't be dispatched again
+            poll(syncingTasks);
+
+            // handle tasks that have failed (i.e., remove them and propagate the errors to the listeners)
+            handleFailures(failedTasks);
+
+            long elapsedTimeNs = System.nanoTime() - cycleStartTimeNs;
+            MicroMeterUtils.time(Duration.ofNanos(elapsedTimeNs), "schedule.cycle");
+            long nextSchedule = Math.max(0, pollPeriod.toNanos() - elapsedTimeNs);
+            Preconditions.checkState(nextSchedule <= pollPeriod.toNanos());
+
+            scheduler.schedule(this::tick, nextSchedule, TimeUnit.NANOSECONDS);
+        } catch (InterruptedException ie) {
```

Comment:
We already have this exception handled in line 129 (tick() method) - should one of these be removed?

Reply:
Maybe also set a flag `interrupted = true` here and check the flag in the beginning of finally block.

Reply:
Done!